google-cloud-storage/src/test/java/com/google/cloud/storage/StorageImplMockitoTest.java

```diff
@@ -0,0 +1,583 @@
+    initializeService();
+    String fileName = "non_existing_file.txt";
+    try {
+      storage.upload(BLOB_INFO1, Paths.get(fileName));
+      Assert.fail();
+    } catch (IOException e) {
+      assertEquals(NoSuchFileException.class, e.getClass());
+      assertEquals(fileName, e.getMessage());
+    }
+  }
+
+  @Test
+  public void testUploadDirectory() throws IOException {
+    initializeService();
+    Path dir = Files.createTempDirectory("unit_");
+    try {
+      storage.upload(BLOB_INFO1, dir);
+      Assert.fail();
+    } catch (StorageException e) {
+      assertEquals(dir + " is a directory", e.getMessage());
+    }
+  }
+
+  private static class UploadParameters {
+    final String uploadId;
+    final byte[] buffer;
+    final int length;
+    final boolean isLast;
+    final BlobInfo blobInfo;
+
+    private UploadParameters(
+        String uploadId, byte[] buffer, int length, boolean isLast, BlobInfo blobInfo) {
+      this.uploadId = uploadId;
+      this.buffer = buffer;
+      this.length = length;
+      this.isLast = isLast;
+      this.blobInfo = blobInfo;
+    }
+  }
+
+  private UploadParameters initializeUpload(byte[] bytes) {
+    return initializeUpload(bytes, DEFAULT_BUFFER_SIZE, EMPTY_RPC_OPTIONS);
+  }
+
+  private UploadParameters initializeUpload(byte[] bytes, int bufferSize) {
+    return initializeUpload(bytes, bufferSize, EMPTY_RPC_OPTIONS);
+  }
+
+  private UploadParameters initializeUpload(
+      byte[] bytes, int bufferSize, Map<StorageRpc.Option, ?> rpcOptions) {
+    String uploadId = "upload-id";
+    byte[] buffer = new byte[bufferSize];
+    System.arraycopy(bytes, 0, buffer, 0, bytes.length);
+    BlobInfo blobInfo = BLOB_INFO1.toBuilder().setMd5(null).setCrc32c(null).build();
+    when(storageRpcMock.open(blobInfo.toPb(), rpcOptions)).thenReturn(uploadId);
+    initializeService();
+    return new UploadParameters(uploadId, buffer, bytes.length, true, blobInfo);
+  }
+
+  private void verifyUpload(UploadParameters parameters) {
+    verify(storageRpcMock)
+        .write(parameters.uploadId, parameters.buffer, 0, 0L, parameters.length, parameters.isLast);
+  }
+
+  @Test
+  public void testUploadFile() throws Exception {
+    byte[] dataToSend = {1, 2, 3, 4};
+    Path tempFile = Files.createTempFile("testUpload", ".tmp");
+    Files.write(tempFile, dataToSend);
+
+    UploadParameters uploadParameters = initializeUpload(dataToSend);
+    storage.upload(uploadParameters.blobInfo, tempFile);
+    verifyUpload(uploadParameters);
+  }
+
+  @Test
+  public void testUploadStream() throws Exception {
+    byte[] dataToSend = {1, 2, 3, 4, 5};
+    ByteArrayInputStream stream = new ByteArrayInputStream(dataToSend);
+
+    UploadParameters uploadParameters = initializeUpload(dataToSend);
+    storage.upload(uploadParameters.blobInfo, stream);
+    verifyUpload(uploadParameters);
+  }
+
+  @Test
+  public void testUploadWithOptions() throws Exception {
+    byte[] dataToSend = {1, 2, 3, 4, 5, 6};
```

Comment:
is it possible to declare `dataToSend `globally ?

Reply:
dataToSend varies from test to test.